java-frontend/src/main/java/org/sonar/java/se/checks/SECheck.java

```diff
@@ -60,13 +64,50 @@ public void checkEndOfExecutionPath(CheckerContext context, ConstraintManager co
   @Override
   public void scanFile(JavaFileScannerContext context) {
     for (SEIssue seIssue : issues) {
-      context.reportIssue(this, seIssue.getTree(), seIssue.getMessage(), seIssue.getSecondary(), null);
+      context.reportIssueWithFlow(this, seIssue.getTree(), seIssue.getMessage(), seIssue.getFlows(), null);
     }
     issues.clear();
   }
 
-  public void reportIssue(Tree tree, String message, List<JavaFileScannerContext.Location> secondary) {
-    issues.add(new SEIssue(tree, message, secondary));
+  public void reportIssue(Tree tree, String message, Set<List<JavaFileScannerContext.Location>> secondary) {
+    issues.add(issues.stream()
+      .filter(seIssue -> seIssue.tree.equals(tree))
+      .findFirst()
+      .map(seIssue -> {
+        seIssue.flows.addAll(secondary);
+        return seIssue;
+      })
+      .orElse(new SEIssue(tree, message, secondary)));
+  }
+
+  public static List<JavaFileScannerContext.Location> flow(ExplodedGraph.Node currentNode, SymbolicValue currentVal ) {
+    List<JavaFileScannerContext.Location> flow = new ArrayList<>();
+    ExplodedGraph.Node node = currentNode;
+    Symbol lastEvaluated = currentNode.programState.getLastEvaluated();
+    while (node != null) {
+      ExplodedGraph.Node finalNode = node;
+      if(finalNode.programPoint.syntaxTree() != null) {
+        node.learnedConstraints.stream()
+          .map(lc->lc.sv)
+          .filter(sv -> sv.equals(currentVal))
+          .findFirst()
+          .ifPresent(sv -> flow.add(new JavaFileScannerContext.Location("", finalNode.parent.programPoint.syntaxTree())));
+        if (lastEvaluated != null) {
+          Symbol finalLastEvaluated = lastEvaluated;
+          Optional<Symbol> learnedSymbol = node.getLearnedSymbols().stream()
+            .map(ls -> ls.symbol)
+            .filter(sv -> sv.equals(finalLastEvaluated))
+            .findFirst();
+          if (learnedSymbol.isPresent()) {
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Refactor this code to not nest more than 3 if/for/while/switch/try statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS134)